django/db/models/base.py

```diff
@@ -407,91 +407,102 @@ def __init__(self, *args, **kwargs):
         # overrides it. It should be one or the other; don't duplicate the work
         # The reason for the kwargs check is that standard iterator passes in by
         # args, and instantiation for iteration is 33% faster.
-        if len(args) > len(opts.concrete_fields):
-            # Daft, but matches old exception sans the err msg.
-            raise IndexError("Number of args exceeds number of fields")
-
-        if not kwargs:
-            fields_iter = iter(opts.concrete_fields)
-            # The ordering of the zip calls matter - zip throws StopIteration
-            # when an iter throws it. So if the first iter throws it, the second
-            # is *not* consumed. We rely on this, so don't change the order
-            # without changing the logic.
-            for val, field in zip(args, fields_iter):
+        concrete_fields_attnames_enumerated = opts.concrete_fields_attnames_enumerated
+        fields_len = len(concrete_fields_attnames_enumerated)
+        args_len = len(args)
+
+        if args_len == fields_len:
+            for i, attname in concrete_fields_attnames_enumerated:
+                val = args[i]
                 if val is _DEFERRED:
                     continue
-                _setattr(self, field.attname, val)
+                _setattr(self, attname, val)
         else:
-            # Slower, kwargs-ready version.
-            fields_iter = iter(opts.fields)
-            for val, field in zip(args, fields_iter):
-                if val is _DEFERRED:
-                    continue
-                _setattr(self, field.attname, val)
-                kwargs.pop(field.name, None)
+            if args_len > fields_len:
+                # Daft, but matches old exception sans the err msg.
+                raise IndexError("Number of args exceeds number of fields")
+
+            if not kwargs:
+                fields_iter = iter(opts.concrete_fields)
+                # The ordering of the zip calls matter - zip throws StopIteration
+                # when an iter throws it. So if the first iter throws it, the second
+                # is *not* consumed. We rely on this, so don't change the order
+                # without changing the logic.
```

Comment:
Probably , "We rely on this, so don't change the order without changing the logic." can be removed or else just say, "Don't change the order without changing the logic."